app/src/main/java/com/fstyle/structure_android/data/repository/UserRepository.java

```diff
@@ -37,196 +22,101 @@
 public class UserRepository
         implements UserDataSource.RemoteDataSource, UserDataSource.LocalDataSource {
 
-    private static final String SELECT_ALL_USER_QUERY =
-            String.format("SELECT * FROM %s", UserDbHelper.UserEntry.TABLE_NAME);
+    private static UserRepository INSTANCE;
+
+    @NonNull
+    private UserLocalDataSource mUserLocalDataSource;
+
+    @NonNull
+    private UserRemoteDataSource mUserRemoteDataSource;
 
-    private NameApi mNameApi;
-    private SQLiteDatabase mDatabase;
+    /**
+     * This variable has package local visibility so it can be accessed from tests.
+     */
+    @VisibleForTesting
+    @Nullable
+    private List<User> mCachedUsers;
 
-    public UserRepository(SQLiteDatabase database, NameApi api) {
-        mDatabase = database;
-        mNameApi = api;
+    /**
+     * Marks the cache as invalid, to force an update the next time data is requested. This variable
+     * has package local visibility so it can be accessed from tests.
+     */
+    @VisibleForTesting
+    boolean mCacheIsDirty = false;
+
+    public UserRepository(@NonNull UserLocalDataSource userLocalDataSource,
+            @NonNull UserRemoteDataSource userRemoteDataSource) {
+        mUserLocalDataSource = checkNotNull(userLocalDataSource);
+        mUserRemoteDataSource = checkNotNull(userRemoteDataSource);
+    }
+
+    /**
+     * Returns the single instance of this class, creating it if necessary.
+     *
+     * @param remoteDataSource the backend data source
+     * @param localDataSource the device storage data source
+     * @return the {@link UserRepository} instance
+     */
+    public static synchronized UserRepository getInstance(
+            @NonNull UserLocalDataSource localDataSource,
+            @NonNull UserRemoteDataSource remoteDataSource) {
+        if (INSTANCE == null) {
+            INSTANCE = new UserRepository(checkNotNull(localDataSource),
+                    checkNotNull(remoteDataSource));
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Used to force {@link #getInstance(UserLocalDataSource, UserRemoteDataSource)} to create a new
+     * instance
+     * next time it's called.
+     */
+    public static void destroyInstance() {
+        INSTANCE = null;
     }
 
     @Override
-    public Single<List<User>> searchUsers(String keyWord, String limit) {
-        return mNameApi.searchGithubUsers(keyWord, limit)
-                .flatMap(new Function<SearchUserResponse, SingleSource<? extends List<User>>>() {
-                    @Override
-                    public SingleSource<? extends List<User>> apply(
-                            SearchUserResponse searchUserResponse) throws Exception {
-                        return Single.just(searchUserResponse.getItems());
-                    }
-                });
+    public Completable insertListUser(List<User> users) {
+        return mUserLocalDataSource.insertListUser(Preconditions.checkNotNull(users));
```

Comment:
WARNING: [UnnecessaryFullyQualifiedName] Unnecessary use of fully qualified name 'Preconditions.checkNotNull' due to existing static import 'com.google.common.base.Preconditions.checkNotNull'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnnecessaryFullyQualifiedName